### PR TITLE
Fix runout sensor for mixing extruder

### LINKED ISSUE #19877: [BUG] M600 (Filemant Change) not working in filament runout script for mixing extruders
### Bug Description

Normally M600 falls back to the active extrude if no extrude / tool is specified with `Tx`. This works for single and (basic) multiple extruders, however once `MIXING_EXTRUDER` is enabled an *"E stepper not specified"* error is thrown if parameter `Tx` is omitted. 
I would have expected a similar behavior as for a basic dual extruder setup. Especially if you are using `M600` as the filament runout script.

#### dual extrude setup -> OK

``` C
#define EXTRUDERS 2
#define SINGLENOZZLE
// #define MIXING_EXTRUDER
#define NUM_RUNOUT_SENSORS   2
#define FILAMENT_RUNOUT_SCRIPT "M600"
```

#### mixing extrude setup -> *"E stepper not specified"* error

``` C
#define EXTRUDERS 1
//#define SINGLENOZZLE
#define MIXING_EXTRUDER
#define NUM_RUNOUT_SENSORS   2
#define FILAMENT_RUNOUT_SCRIPT "M600"
```

### Steps to Reproduce

1. start a print
2. trigger the filament runout sensor

(or call `M600` via terminal)

**Expected behavior:**

Execute filament change for active extruder. Or even better, execute filament change for *triggering extruder*.

**Actual behavior:**

*"E stepper not specified"* error and print continues (without filament).

### Additional Information

I traced this down to different error handling (missing `Tx`) of the functions `get_target_extruder_from_command()` (used with dual extruders) and `get_target_e_stepper_from_command()` (used with mixing extruders). Is this intended?

Maybe it would be even better to be able to pass the triggering filament sensor to the runout script. e.g. `"M600 T%t"`

### PR DESCRIPTION
**Pull request by the-real-orca**
### Description

Fixing runout sensor to use actually triggering extruder / sensor. The actual (virtual) extruder cannot be used with mixing setup as it is not necessarily mapping to a single extruder.

The triggering sensor can now be used as parameter (_%t_) for the `FILAMENT_RUNOUT_SCRIPT`.
e.g. `"M600 T%t"`

### Benefits

Enables runout sensors and runout script for mixing extrude setup.

### Configurations

Configuration.h / `FILAMENT_RUNOUT_SENSOR` section:

`WATCH_ALL_RUNOUT_SENSORS`: Triggers runout script for all sensors (not only the one for the active sensor). This is automatically enabled for `MIXING_EXTRUDER`.

### Related Issues

Related to bug report #19877

### CODE REVIEW

File: Marlin/src/feature/runout.cpp [hunk 1]
```diff
@@ -84,15 +84,11 @@ void event_filament_runout() {
...
-      #if NUM_RUNOUT_SENSORS > 1
-        + active_extruder
-      #endif
-    ;
-  #endif
+  const char tool = '0'
+    #if E_STEPPERS > 1
+      + extruder
+    #endif
+  ;
```
Comment (sjasonsmith) [inline Marlin/src/feature/runout.cpp]: You just dropped the ExtUI code?

File: Marlin/src/feature/runout.h [hunk 2]
```diff
@@ -285,16 +315,21 @@ class FilamentSensorBase {
...
-          static bool was_out = false;
-          if (out != was_out) {
-            was_out = out;
-            SERIAL_ECHOPGM("Filament ");
-            serialprintPGM(out ? PSTR("OUT\n") : PSTR("IN\n"));
-          }
+        #if ENABLED(FILAMENT_RUNOUT_SENSOR_DEBUG)
+          static bool was_out[NUM_RUNOUT_SENSORS] = { false };
         #endif
+        LOOP_L_N(s, NUM_RUNOUT_SENSORS) {
```
Comment (sjasonsmith) [inline Marlin/src/feature/runout.h:313]: Is this looking a all sensors even when `WATCH_ALL_RUNOUT_SENSORS` is not enabled?

File: Marlin/src/feature/runout.h [hunk 3]
```diff
@@ -334,8 +369,10 @@ class FilamentSensorBase {
         #endif
       }
 
-      static inline bool has_run_out() {
-        return runout_mm_countdown[active_extruder] < 0;
+      static inline uint8_t has_run_out() {
+        uint8_t sensor_bitmask = 0;
+        LOOP_L_N(i, NUM_RUNOUT_SENSORS) if (runout_mm_countdown[i] < 0) SBI(sensor_bitmask, i);
```
Comment (sjasonsmith) [inline Marlin/src/feature/runout.h]: Also seems to ignore whether your feature is enabled?

Comment (sjasonsmith) [reply Marlin/src/feature/runout.h]: This applies to everything below this in the file also.

File: Marlin/src/inc/SanityCheck.h [hunk 4]
```diff
@@ -811,12 +811,10 @@ static_assert(Y_MAX_LENGTH >= Y_BED_SIZE, "Movement bounds (Y_MIN_POS, Y_MAX_POS
...
     #error "FILAMENT_RUNOUT_SENSOR requires FIL_RUNOUT_PIN."
-  #elif NUM_RUNOUT_SENSORS > E_STEPPERS
-    #if HAS_PRUSA_MMU2
+  #elif HAS_PRUSA_MMU2 && NUM_RUNOUT_SENSORS != 1
       #error "NUM_RUNOUT_SENSORS must be 1 with MMU2 / MMU2S."
-    #else
-      #error "NUM_RUNOUT_SENSORS cannot exceed the number of E steppers."
-    #endif
+  #elif !(NUM_RUNOUT_SENSORS == 1 || NUM_RUNOUT_SENSORS == E_STEPPERS)
+    #error "NUM_RUNOUT_SENSORS must be either 1 or number of E steppers."
```
Comment (sjasonsmith) [inline Marlin/src/inc/SanityCheck.h]: Is this an existing limitation or a new one?

I assumed it was possible to have two extruders but only have a filament sensor on one of them.

File: Marlin/src/feature/runout.cpp (hunk 1 above)
Comment (the-real-orca) [reply Marlin/src/feature/runout.cpp]: True, that was a mistake. It’s fixed in the last commit.

File: Marlin/src/feature/runout.h (hunk 2 above)
Comment (the-real-orca) [reply Marlin/src/feature/runout.h:313]: The state is checked for all sensors, however a runout is only triggered for the active extruder / sensor or when WATCH_ALL_RUNOUT_SENSORS is enabled.

File: Marlin/src/feature/runout.h (hunk 3 above)
Comment (the-real-orca) [reply Marlin/src/feature/runout.h]: The state is checked for all sensors, however a runout is only triggered for the active extruder / sensor or when WATCH_ALL_RUNOUT_SENSORS is enabled.

File: Marlin/src/inc/SanityCheck.h (hunk 4 above)
Comment (the-real-orca) [reply Marlin/src/inc/SanityCheck.h]: The logic has changed a bit, but the general limitation was already there before.
You can have one sensor on one extruder or sensors on all extruders (or multiple sensors wired parallel to one sensor input).
The change only affects you if you have 3 or more extruders. Here is the question how realistic is it to have sensors on only 2 out of 3 extruders vs. making a configuration mistake.

File: Marlin/src/lcd/extui/ui_api.cpp [hunk 5]
```diff
@@ -340,8 +340,10 @@ namespace ExtUI {
     #endif
   }
 
-  extruder_t getActiveTool() {
-    switch (active_extruder) {
+  extruder_t getTool(const uint8_t extruder) {
+    switch (extruder) {
+      case 7:  return E6;
+      case 6:  return E7;
```
Comment (fishcu) [inline Marlin/src/lcd/extui/ui_api.cpp]: Seems like E6 and E7 got switched here?

Comment (the-real-orca) [reply Marlin/src/lcd/extui/ui_api.cpp]: Thanks for noticing this glitch. Just fixed (cd59fa51000588338fc367c0c6f1d4c31b7da7ef).

File: Marlin/src/inc/SanityCheck.h (hunk 4 above)
Comment (thinkyhead) [reply Marlin/src/inc/SanityCheck.h]: Good question. Still, we allow for it because… maybe the third extruder is melting chocolate chips…?

**Comment by tommylux** [comment]
Is the runout sensor issue now fixed for the likes of A20T?
At least my last bugfix firmware build is 2022-01-17, and at lease my E2 (extruder 3) sensor is not working.
Incidentally, I have the config set to:
  #define NUM_RUNOUT_SENSORS   3          // Number of sensors, up to one per extruder. Define a FIL_RUNOUT#_PIN for each.
  #define FIL_RUNOUT_PIN      66
  #define FIL_RUNOUT2_PIN     67
  #define FIL_RUNOUT3_PIN     54  // TL apparently 54 not 68 according to https://github.com/MarlinFirmware/Marlin/issues/19877#issuecomment-742643079

The runout sensor didn't work.

**Pull request merged**